File: cpp_sdk/cpp_dist_bundle/c++/detect_eyeglasses.cpp
// Sample code: load an image, detect the largest face and check whether the face is wearing eye glasses or not
// First image is of a person wearing eye glasses. Second image is of a person not wearing eye glasses.
// The probability that eye glasses of some type are worn over the face is computed in both cases.

#include "tf_sdk.h"
#include <iostream>

int main() {
    Trueface::SDK tfSdk;
    // TODO: Either input your token in the CMakeLists.txt file, or insert it below directly
    bool valid = tfSdk.setLicense(TRUEFACE_TOKEN);

    if (!valid) {
        std::cout<<"Error: the provided license is invalid."<<std::endl;
        return 1;
    }

    // Load the glasses image and detect largest face.
    std::cout << "Image with glasses:" << std::endl;
    Trueface::ErrorCode errorCode = tfSdk.setImage("../../images/glasses.jpg");
    if (errorCode != Trueface::ErrorCode::NO_ERROR) {
        std::cout<<"Error: could not load the image"<<std::endl;
        return 1;
    }

    Trueface::FaceBoxAndLandmarks faceBoxAndLandmarks;
    bool found = false;
    errorCode = tfSdk.detectLargestFace(faceBoxAndLandmarks, found);

    if (!found || errorCode != Trueface::ErrorCode::NO_ERROR) {
        std::cout<<"Error: face not found";
        return 1;
    }

    // Run glasses detection
    Trueface::GlassesLabel glassesLabel;
    float glassesScore;
    errorCode = tfSdk.detectGlasses(faceBoxAndLandmarks, glassesLabel, glassesScore);

    if (errorCode != Trueface::ErrorCode::NO_ERROR) {
        std::cout<<"Error: could not run glasses detection"<<std::endl;
        return 1;
    }

    if (glassesLabel == Trueface::GlassesLabel::GLASSES) {
        std::cout << "Glasses detected" << std::endl;
    } else {
        std::cout << "No glasses detected" << std::endl;
    }

    // Load the non glasses image and detect largest face.
    std::cout << "Image with glasses" << std::endl;
    errorCode = tfSdk.setImage("../../images/headshot.jpg");
    if (errorCode != Trueface::ErrorCode::NO_ERROR) {
        std::cout<<"Error: could not load the image"<<std::endl;
        return 1;
    }

    errorCode = tfSdk.detectLargestFace(faceBoxAndLandmarks, found);

    if (!found || errorCode != Trueface::ErrorCode::NO_ERROR) {
        std::cout<<"Error: face not found";
        return 1;
    }

    // Run glasses detection
    errorCode = tfSdk.detectGlasses(faceBoxAndLandmarks, glassesLabel, glassesScore);

    if (errorCode != Trueface::ErrorCode::NO_ERROR) {
        std::cout<<"Error: could not run glasses detection"<<std::endl;
        return 1;
    }

    if (glassesLabel == Trueface::GlassesLabel::GLASSES) {
        std::cout << "Glasses detected" << std::endl;
    } else {
        std::cout << "No glasses detected" << std::endl;
    }
}

File: python/python_dist_bundle/detect_eyeglasses.py
# Sample code: Run eye glasses detection on two images (one image with eye glasses, one image without eye glasses)

import tfsdk
import os
from colorama import Fore
from colorama import Style

options = tfsdk.ConfigurationOptions()
# Can set configuration options here
# ex:
# options.smallest_face_height = 40
# options.fd_filter = tfsdk.FACEDETECTIONFILTER.BALANCED
# options.fr_model = tfsdk.FACIALRECOGNITIONMODEL.TFV5
# Note, if you do use TFV5, you will need to run the download script in /download_models to obtain the model file

sdk = tfsdk.SDK(options)

# TODO: export your license token as TRUEFACE_TOKEN environment variable
is_valid = sdk.set_license(os.environ['TRUEFACE_TOKEN'])
if (is_valid == False):
    print(f"{Fore.RED}Invalid License Provided{Style.RESET_ALL}")
    print(f"{Fore.RED}Be sure to export your license token as TRUEFACE_TOKEN{Style.RESET_ALL}")
    quit()

# Load the image with glasses
res = sdk.set_image("../images/glasses.jpg")
if (res != tfsdk.ERRORCODE.NO_ERROR):
    print(f"{Fore.RED}Unable to set image 1{Style.RESET_ALL}")
    quit()


# detect the largest face
found, face_bounding_box = sdk.detect_largest_face()
if found:
    # Run glasses detection
    res, glasses_label, glasses_score = sdk.detect_glasses(face_bounding_box)
    if (res != tfsdk.ERRORCODE.NO_ERROR):
        print(f"{Fore.RED}Unable to run glasses detection with image 1{Style.RESET_ALL}")
        quit()

    # Expect the result to be "glasses"
    print(f"Face with glasses, predicted result: {glasses_label}")

else:
    print("No face detected in image")

# Load the non glasses face image
res = sdk.set_image("../images/headshot.jpg")
if (res != tfsdk.ERRORCODE.NO_ERROR):
    print(f"{Fore.RED}Unable to set image 2{Style.RESET_ALL}")
    quit()

# detect the largest face
found, face_bounding_box = sdk.detect_largest_face()
if found:
    # Run glasses detection
    res, glasses_label, glasses_score = sdk.detect_glasses(face_bounding_box)
    if (res != tfsdk.ERRORCODE.NO_ERROR):
        print(f"{Fore.RED}Unable to run glasses detection with image 2{Style.RESET_ALL}")
        quit()


    # Expect the result to be "no glasses"
    print(f"Face without glasses, predicted result: {glasses_label}")

else:
    print("No face detected in image")
